lib/srv/alpnproxy/local_proxy.go

```diff
@@ -129,7 +140,14 @@ func (l *LocalProxy) Start(ctx context.Context) error {
 			log.WithError(err).Errorf("Failed to accept client connection.")
 			return trace.Wrap(err)
 		}
+
 		go func() {
+			if l.cfg.Middleware != nil {
+				if err := l.cfg.Middleware.OnNewConnection(ctx, l, conn); err != nil {
+					log.WithError(err).Errorf("Failed to handle connection.")
+					return
+				}
+			}
```

Comment:
I recall one discussion we had is on whether should the callback blocks or not. Some database clients may create multiple connections at the same time. I don't think our middleware is thread-safe (see #5577).

@ravicious to confirm whether Connect prefers the callback to block or not.

Alternatively, if we want both sync/async, maybe we can have both in the interface.

Reply:
I preferred the async method before because I wasn't just checking if the certs expired - I was checking if MFA was required for each connection and treating each connection as a new session, so it made more sense back then. I just didn't change it after we decided to only check cert expiration, and I think the UX is better with a blocking call, so I will change this to block before spawning the goroutine.

Regarding thread-safety: I don't think the async version has any undefined behavior since certs are guarded by a mutex and copied into the ALPN dialer. Suppose you have goroutines `A` and `B`: `A` checks certs and decides they don't need to be renewed, so it proceeds to handle the downstream connection. Before `A` configures its dialer, `B` checks certs and decides they need to be renewed, and then renews the certs. Now `A` and `B` will use the latest certs for their dialers.

Reply:
Sorry, I didn't have time to do an actual review of the PR so I'm just responding to the mention.

> @ravicious to confirm whether Connect prefers the callback to block or not.

I'm still new to the Go world, but I think the callback can block and then the caller can decide to create a goroutine within the callback if they want the callback to become essentially nonblocking?

For a notification about an expired cert when the user attempts to connect to the db, tsh daemon is going to make a gRPC call to the server running in the Electron app. This can be completely async.

For per-session MFA in db tunnels, I haven't looked into that closely. But I imagine at some point tshd will wait for Connect to solve the MFA challenge. I'm not sure yet if this will block and the connection won't be created until the challenge is solved. But I imagine this is how it works in tsh right now with per-session MFA for stuff such as `tsh ssh`.

I'm probably of not much help here until I dive into the rest of the PR.

Reply:
Changed this to a blocking check before the goroutine, but I'll leave this convo unresolved if we want to discuss moving the code in the goroutine into the interface.